File: AndroidStudio_Android/MobileSeconndHand/app/src/main/java/marcin_szyszka/mobileseconndhand/services/GpsLocationService.java
package marcin_szyszka.mobileseconndhand.services;

import android.Manifest;
import android.app.Service;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.IBinder;
import android.provider.Settings;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AlertDialog;
import android.util.Log;

import marcin_szyszka.mobileseconndhand.R;
import marcin_szyszka.mobileseconndhand.models.CoordinatesModel;

/**
 * Created by marcianno on 2016-03-07.
 */
/*public class GpsLocationService {
    private static GpsLocationService ourInstance = new GpsLocationService();

    public static GpsLocationService getInstance() {
        return ourInstance;
    }


}*/

public class GpsLocationService extends Service implements LocationListener {

    private final Context mContext;

    // Flag for GPS status
    boolean isGPSEnabled = false;

    // Flag for network status
    boolean isNetworkEnabled = false;

    // Flag for GPS status
    boolean canGetLocation = false;

    Location location; // Location
    double latitude; // Latitude
    double longitude; // Longitude

    // The minimum distance to change Updates in meters
    private static final long MIN_DISTANCE_CHANGE_FOR_UPDATES = 10; // 10 meters

    // The minimum time between updates in milliseconds
    private static final long MIN_TIME_BW_UPDATES = 1000 * 60 * 1; // 1 minute

    // Declaring a Location Manager
    protected LocationManager locationManager;

    public GpsLocationService(Context context) {
        this.mContext = context;
        this.location = getLocation();
    }

    public Location getLocation() {
        try {
            locationManager = (LocationManager) mContext
                    .getSystemService(LOCATION_SERVICE);

            // Getting GPS status
            isGPSEnabled = locationManager
                    .isProviderEnabled(LocationManager.GPS_PROVIDER);

            // Getting network status
            isNetworkEnabled = locationManager
                    .isProviderEnabled(LocationManager.NETWORK_PROVIDER);

            if (!isGPSEnabled && !isNetworkEnabled) {
                // No network provider is enabled
            } else {
                this.canGetLocation = true;
                if (isNetworkEnabled) {
                    if (ActivityCompat.checkSelfPermission(this.mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this.mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                    }
                    locationManager.requestLocationUpdates(
                            LocationManager.NETWORK_PROVIDER,
                            MIN_TIME_BW_UPDATES,
                            MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
                    if (locationManager != null) {
                        location = locationManager
                                .getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
                        if (location != null) {
                            latitude = location.getLatitude();
                            longitude = location.getLongitude();
                        }
                    }
                }
                // If GPS enabled, get latitude/longitude using GPS Services
                if (isGPSEnabled) {
                    if (location == null) {
                        locationManager.requestLocationUpdates(
                                LocationManager.GPS_PROVIDER,
                                MIN_TIME_BW_UPDATES,
                                MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
                        Log.d("GPS Enabled", "GPS Enabled");
                        if (locationManager != null) {
                            location = locationManager
                                    .getLastKnownLocation(LocationManager.GPS_PROVIDER);
                            if (location != null) {
                                latitude = location.getLatitude();
                                longitude = location.getLongitude();
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return location;
    }


    /**
     * Stop using GPS listener
     * Calling this function will stop using GPS in your app.
     */
    public void stopUsingGPS() {
        if (locationManager != null) {
            if (ActivityCompat.checkSelfPermission(this.mContext, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this.mContext, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                // TODO: Consider calling
                //    ActivityCompat#requestPermissions
                // here to request the missing permissions, and then overriding
                //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
                //                                          int[] grantResults)
                // to handle the case where the user grants the permission. See the documentation
                // for ActivityCompat#requestPermissions for more details.
                return;
            }
            locationManager.removeUpdates(this);
        }
    }


    /**
     * Function to get latitude
     */
    public double getLatitude() {
        if (location != null) {
            latitude = location.getLatitude();
        }

        // return latitude
        return latitude;
    }


    /**
     * Function to get longitude
     */
    public double getLongitude() {
        if (location != null) {
            longitude = location.getLongitude();
        }

        // return longitude
        return longitude;
    }

    /**
     * Function to check GPS/Wi-Fi enabled
     *
     * @return boolean
     */
    public boolean canGetLocation() {
        return this.canGetLocation;
    }


    /**
     * Function to show settings alert dialog.
     * On pressing the Settings button it will launch Settings Options.
     */
    public void showSettingsAlert() {
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(mContext);

        // Setting Dialog Title
        alertDialog.setTitle("Ustawienia lokalizacji");

        // Setting Dialog Message
        alertDialog.setMessage("GPS jest wyłączony. Czy chcesz teraz przejść do ustawień aby włączyć lokalizację?");

        // On pressing the Settings button.
        alertDialog.setPositiveButton("Settings", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                mContext.startActivity(intent);
            }
        });

        // On pressing the cancel button
        alertDialog.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });

        // Showing Alert Message
        alertDialog.show();
    }


    @Override
    public void onLocationChanged(Location location) {
    }


    @Override
    public void onProviderDisabled(String provider) {
    }


    @Override
    public void onProviderEnabled(String provider) {
    }


    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {
    }


    @Override
    public IBinder onBind(Intent arg0) {
        return null;
    }

    public CoordinatesModel getCoordinatesModel() {
        CoordinatesModel coordinatesModel = new CoordinatesModel();
        double latitude = getLatitude();
        double longitude = getLongitude();
        if (latitude == 0 || longitude == 0) {
            String homeLatitude = SharedPreferencesService.getInstance().getSpecificSharedPreferenceString(mContext, R.string.homeLatitutde);
            String homeLongitude = SharedPreferencesService.getInstance().getSpecificSharedPreferenceString(mContext, R.string.homeLongitude);
            if (homeLatitude != null && homeLongitude != null){
                latitude = Double.parseDouble(homeLatitude);
                longitude = Double.parseDouble(homeLongitude);
                showToast("Nie mogę ustalić aktualnej lokalizacji. Użyję lokalizacji domowej");
            }
            else{
                //in future ask user for him home location and show activity to set that
                showToast("Nie mogę ustalić aktualnej lokalizacji i nie masz zapisanej lokalizacji domowej.");
            }
        }
        coordinatesModel.Latitude = latitude;
        coordinatesModel.Longitude = longitude;
        coordinatesModel.MaxDistance = 10;
//SharedPreferencesService.getInstance().getSpecificSharedPreferenceInt(mContext, R.string.maxDistanceToAdvertisements);
        return coordinatesModel;
    }

    private void showToast(String message){
        ToastService.getInstance().showToast(mContext, message);
    }
}

File: Xamarin_Android/MobileSecondHand/MobileSecondHand.App/Infrastructure/GpsLocationService.cs
using System;
using Android.App;
using Android.Content;
using Android.Locations;
using Android.OS;
using Android.Provider;
using Android.Runtime;
using MobileSecondHand.API.Models.Shared.Consts;
using MobileSecondHand.API.Models.Shared.Location;
using MobileSecondHand.App.Consts;
using MobileSecondHand.Models.Consts;
using MobileSecondHand.Models.Settings;
using MobileSecondHand.Services.Location;

namespace MobileSecondHand.App.Infrastructure
{
	public class GpsLocationService : Service, ILocationListener, IDialogInterfaceOnClickListener
	{
		private bool isGPSEnabled;
		private bool isNetworkEnabled;
		private double latitude;
		private double longitude;
		private Location location;
		private LocationManager locationManager;
		private Context mContext;
		SharedPreferencesHelper sharedPreferencesHelper;
		ISettingWindowCloseListener settingWindowListener;
		private long MIN_DISTANCE_CHANGE_FOR_UPDATES = 2000;//500m
		static GpsLocationService serviceInstance;

		private long MIN_TIME_BW_UPDATES = 1000 * 60 * 1 * 60; // 1 hour

		public bool CanGetLocation
		{
			get
			{
				// Getting GPS status
				if (!locationManager.IsProviderEnabled(LocationManager.GpsProvider) && !locationManager.IsProviderEnabled(LocationManager.NetworkProvider))
				{
					return false;
				}
				else
				{
					return true;
				}
			}
		}

		public static GpsLocationService GetServiceInstance(Context context, ISettingWindowCloseListener listener = null)
		{
			if (serviceInstance == null)
			{
				serviceInstance = new GpsLocationService(context, listener);
			}

			return serviceInstance;
		}

		public GpsLocationService(Context context, ISettingWindowCloseListener settingWindowListener = null)
		{
			this.mContext = context;
			this.settingWindowListener = settingWindowListener;
			this.sharedPreferencesHelper = new SharedPreferencesHelper(context);
			this.location = GetLocation();
		}

		public Location GetLocation()
		{
			try
			{
				locationManager = (LocationManager)mContext.GetSystemService("location");

				// Getting GPS status
				isGPSEnabled = locationManager.IsProviderEnabled(LocationManager.GpsProvider);

				// Getting network status
				isNetworkEnabled = locationManager.IsProviderEnabled(LocationManager.NetworkProvider);

				if (!isGPSEnabled && !isNetworkEnabled)
				{
					//showalert in activity
				}
				else
				{
					if (isNetworkEnabled)
					{
						if (locationManager != null)
						{
							locationManager.RequestLocationUpdates(
									LocationManager.NetworkProvider,
									MIN_TIME_BW_UPDATES,
									MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
							//locationManager.RequestSingleUpdate(LocationManager.NetworkProvider, this, null);
							location = locationManager.GetLastKnownLocation(LocationManager.NetworkProvider);
							if (location != null)
							{
								latitude = location.Latitude;
								longitude = location.Longitude;
							}
						}
					}
					if (isGPSEnabled)
					{
						if (location == null)
						{
							locationManager.RequestLocationUpdates(
									LocationManager.GpsProvider,
									MIN_TIME_BW_UPDATES,
									MIN_DISTANCE_CHANGE_FOR_UPDATES, this);
							//locationManager.RequestSingleUpdate(LocationManager.GpsProvider, this, null);
							if (locationManager != null)
							{
								location = locationManager.GetLastKnownLocation(LocationManager.GpsProvider);
								if (location != null)
								{
									latitude = location.Latitude;
									longitude = location.Longitude;
								}
							}
						}
					}
				}
			}
			catch (Exception e)
			{
				AlertsService.ShowLongToast(this.mContext, "Nie masz w³¹czonej lokalizacji!");
			}

			return location;
		}

		public CoordinatesForAdvertisementsModel GetCoordinatesModel(int maxDistance = -1)
		{
			CoordinatesForAdvertisementsModel coordinatesModel = new CoordinatesForAdvertisementsModel();
			var location = GetLocation();
			var settingsMOdel = (AppSettingsModel)sharedPreferencesHelper.GetSharedPreference<AppSettingsModel>(SharedPreferencesKeys.APP_SETTINGS);
			if (location == null || latitude == 0 || longitude == 0)
			{
				AlertsService.ShowLongToast(this.mContext, "Nie mogê ustaliæ aktualnej lokalizacji. SprawdŸ ustawienia GPS");
				throw new Exception("Brak lokalizacji");
			}
			else
			{
				coordinatesModel.Latitude = latitude;
				coordinatesModel.Longitude = longitude;
				if (maxDistance > -1)
				{
					coordinatesModel.MaxDistance = maxDistance;
				}
				else if (settingsMOdel != null)
				{
					coordinatesModel.MaxDistance = settingsMOdel.LocationSettings.MaxDistance;
				}
				else
				{
					coordinatesModel.MaxDistance = ValueConsts.MAX_DISTANCE_VALUE;
				}
			}

			return coordinatesModel;
		}

		public void ShowSettingsAlert()
		{
			AlertDialog.Builder alertDialog = new AlertDialog.Builder(mContext);
			alertDialog.SetTitle("Ustawienia lokalizacji");
			alertDialog.SetMessage("GPS jest wy³¹czony. Czy chcesz teraz przejœæ do ustawieñ aby w³¹czyæ lokalizacjê?");

			alertDialog.SetPositiveButton("PrzejdŸ do ustawieñ", this);
			alertDialog.SetNegativeButton("Anuluj", this);
			alertDialog.SetCancelable(false);
			alertDialog.Show();
		}


		public void OnClick(IDialogInterface dialog, int which)
		{
			if (which == -1)
			{
				//ok
				Intent intent = new Intent(Settings.ActionLocationSourceSettings);
				mContext.StartActivity(intent);
			}
			else
			{
				settingWindowListener.OnSettingsWindowClose();
			}

		}

		public override IBinder OnBind(Intent intent)
		{
			return null;
		}

		public void OnLocationChanged(Location location)
		{
			//throw new NotImplementedException();
		}

		public void OnProviderDisabled(string provider)
		{
			//throw new NotImplementedException();
		}

		public void OnProviderEnabled(string provider)
		{
			//throw new NotImplementedException();
		}

		public void OnStatusChanged(string provider, [GeneratedEnum] Availability status, Bundle extras)
		{
			//throw new NotImplementedException();
		}
	}


}